### PR TITLE
Expanded version of the GooglePlaces API

### PR DESCRIPTION
**Pull request by MartijnVanAndel**
Implementation of multiple detail levels: Basis, Contact, Atmosphere and Full, for cheaper requests.
Implementation of a function to retrieve images.
Implementation to use other parameters: sessiontoken, origin and location.
Implementation to retrieve other fields of place: plus_code, url, permanently_closed, international_phone_number, formatted_phone_number, opening_hours, website, price_level, rating, reviews and user_ratings_total.
Auto reset sessiontoken when DetailLevel is used (only if same instance is used).

Possible improvements:

Maybe implement a way to not use a SessionToken at all (now one is generated if null)
Find a way to use the parameters 'offset' (null value turn to 0 and has effect on search behaviour),
'radius' and 'strict_bounds' (these are normally only used in combination with 'location' or 'origin'.

### CODE REVIEW

File: src/GooglePlacesApi.Abstractions/Interfaces/IGooglePlacesApi.cs [hunk 1]
```diff
@@ -9,7 +10,10 @@ public interface IGooglePlacesApi
         [Get("/maps/api/place/autocomplete/json?input={input}")]
         Task<Predictions> GetAutocompleteAsync(string input, GoogleApiQueryStringParameters queryStringParameters);
 
-        [Get("/maps/api/place/details/json?key={apiKey}&placeid={placeId}")]
-        Task<Details> GetDetailsAsync(string apiKey, string placeId);
+        [Get("/maps/api/place/details/json?key={apiKey}&placeid={placeId}&fields={fields}")]
+        Task<Details> GetDetailsAsync(string apiKey, string placeId, string sessionToken, string fields);
+
+        [Get("/maps/api/place/photo?maxwidth=400&photoreference={photoReference}&key={apiKey}")]
```
Comment (jacobduijzer) [inline src/GooglePlacesApi.Abstractions/Interfaces/IGooglePlacesApi.cs]: Is there a reason to use a maxwith of 400? Would it not be better to make it an (optional) parameter?

File: src/GooglePlacesApi.Abstractions/Models/GoogleApi/Place.cs [hunk 2]
```diff
@@ -1,53 +1,90 @@
...
-
-        [JsonProperty("vicinity")]
-        public string Vicinity { get; set; }
-    }
-}
+        [JsonProperty("photos")]
+        public List<Photo> Photos { get; set; }
+
+        [JsonProperty("plus_code")]
+        public PlusCode Pluscode { get; set; }
```
Comment (jacobduijzer) [inline src/GooglePlacesApi.Abstractions/Models/GoogleApi/Place.cs]: PErsonally I would rename it to PlusCode...

File: src/GooglePlacesApi.Abstractions/Models/GoogleApiQueryStringParameters.cs [hunk 3]
```diff
@@ -15,5 +16,23 @@ public class GoogleApiQueryStringParameters
 
         [AliasAs("types")]
         public string PlaceType { get; set; }
+
+        [AliasAs("sessiontoken")]
+        public string SessionToken { get; set; }
+
+        //[AliasAs("offset")]
```
Comment (jacobduijzer) [inline src/GooglePlacesApi.Abstractions/Models/GoogleApiQueryStringParameters.cs]: Remove comment :)

Comment (jacobduijzer) [reply src/GooglePlacesApi.Abstractions/Models/GoogleApiQueryStringParameters.cs]: If not needed, remove the other comments below too

File: src/GooglePlacesApi/Services/GooglePlacesApiService.cs [hunk 4]
```diff
@@ -40,15 +49,57 @@ public async Task<Predictions> GetPredictionsAsync(string searchText)
...
-                             .ConfigureAwait(false);
-        }
-    }
-}
+            string fields = null;
+
+            switch (detailLevel)
+            {
+                case DetailLevel.Basic:
+                    fields = "id,place_id,name,geometry,formatted_address,reference,icon,address_components,adr_address,types,photos,plus_code,scope,url,utc_offset,vicinity,permanently_closed";
```
Comment (jacobduijzer) [inline src/GooglePlacesApi/Services/GooglePlacesApiService.cs]: Maybe refactor this to a slightly different format?

basic = "id,place,blablabla";

case DetailLevel.Contact:
fields = $"{basic},extra1,extra2";
break;

case DetailLevel.Atmosphere:
fields = $"{basic},extra3,extra4}

maybe also store the (magic) strings in the Constants file.

Comment (jacobduijzer) [reply src/GooglePlacesApi/Services/GooglePlacesApiService.cs]: in peudo-code, something like this:

basic = Constants.BasicRequestFields;
contact = $"{Constants.BasicRequestFields},{Constants.ContactRequestFields}";

File: src/GooglePlacesApi/Services/GooglePlacesApiService.cs [hunk 5]
```diff
@@ -40,15 +49,57 @@ public async Task<Predictions> GetPredictionsAsync(string searchText)
...
                                  .ConfigureAwait(false);
         }
-
-
-        public async Task<Details> GetDetailsAsync(string placeId)
-        {
+
+        public async Task<Details> GetDetailsAsync(string placeId, string sessionToken, DetailLevel detailLevel = DetailLevel.Basic)
+        {
             if (string.IsNullOrWhiteSpace(placeId))
```
Comment (jacobduijzer) [inline src/GooglePlacesApi/Services/GooglePlacesApiService.cs:55]: Shouldn't there also be a check on sessionToken? Maybe if it is not null or empty? Or does it work even when sessionToken is empty?

**Review by jacobduijzer — COMMENTED** [review]
Thanks for the RP! I reviewed the code and added some comments. Can you review these? Overall, everything looks good, clean & clear!

File: src/GooglePlacesApi.Abstractions/Interfaces/IGooglePlacesApi.cs (hunk 1 above)
Comment (MartijnVanAndel) [reply src/GooglePlacesApi.Abstractions/Interfaces/IGooglePlacesApi.cs]: True, forgot to look the width and height parameters. I'll implement them both with the default value of 1600 (is the maximum)

File: src/GooglePlacesApi.Abstractions/Models/GoogleApi/Place.cs (hunk 2 above)
Comment (MartijnVanAndel) [reply src/GooglePlacesApi.Abstractions/Models/GoogleApi/Place.cs]: I like that better to: I've changed it.

File: src/GooglePlacesApi.Abstractions/Models/GoogleApiQueryStringParameters.cs (hunk 3 above)
Comment (MartijnVanAndel) [reply src/GooglePlacesApi.Abstractions/Models/GoogleApiQueryStringParameters.cs]: This is a more complicated one: the api allows empty strings in the request. Though _offset_, _radius_ and _strictbounds_ cannot be empty. _Offset_ is optional (a different search method). _Radius_ and _strictbounds_ are conditional: they should only be used in combination with _Location_ or _Origin_. 

I've implemented all code to be able to use the  _offset_, _radius_ and _strictbounds_ parameters, though I've commented them out, to keep them out of the API request. Do you have a suggestion how to implement these parameters conditionally?

File: src/GooglePlacesApi/Services/GooglePlacesApiService.cs (hunk 4 above)
Comment (MartijnVanAndel) [reply src/GooglePlacesApi/Services/GooglePlacesApiService.cs]: Done, stored the field strings in the Constants file, then combining them in the GooglePlacesApiService file.

File: src/GooglePlacesApi/Services/GooglePlacesApiService.cs (hunk 5 above)
Comment (MartijnVanAndel) [reply src/GooglePlacesApi/Services/GooglePlacesApiService.cs:55]: You also get results if the sessionToken is provided in the request like:
...[previous argument]&**sessiontoken=**&[next argument]...

Though I've implemented to always set a sessionToken in the service initialization:
```
   public GooglePlacesApiService(GoogleApiSettings settings) 
        {
        ...
           if (string.IsNullOrEmpty(settings.SessionToken))
                settings.SessionToken = Guid.NewGuid().ToString();
```
Therefore a sessionToken will always be provided.
There is a risk that you use this sessionToken too long, if you use the same initialized service for more then several minutes: in that case you may be billed individually per request by Google.

**Comment by MartijnVanAndel** [comment]
> Thanks for the RP! I reviewed the code and added some comments. Can you review these? Overall, everything looks good, clean & clear!

Hi Jacob, i've committed part of your suggestions. Two items remain open:

1. Do we want to create the possibility to use the _offset_, _radius_ and _strictbounds_ parameters, or leave them out for now? (I don't need them, but adding them would mean a 100% support for the API options)
2. Do we want a cleaner way to (re)set a sessionToken (current solution works, but sessionToken may be used too long if service instance is active for more then just a few minutes)

File: src/GooglePlacesApi.Abstractions/Models/GoogleApiQueryStringParameters.cs (hunk 3 above)
Comment (jacobduijzer) [reply src/GooglePlacesApi.Abstractions/Models/GoogleApiQueryStringParameters.cs]: I think I do not get it. I only meant to say: remove the commented code. Leave the attributes as is.

**Comment by jacobduijzer** [comment]
Hi,

I guess we can leave new features for now. If anyone needs them they can add them. Just keep the PR small and leave it like this. Same for the sessionToken, if there are issues later on we can think about extending / fixing.

File: src/GooglePlacesApi.Abstractions/Models/GoogleApiQueryStringParameters.cs (hunk 3 above)
Comment (MartijnVanAndel) [reply src/GooglePlacesApi.Abstractions/Models/GoogleApiQueryStringParameters.cs]: I've removed the comments

**Comment by MartijnVanAndel** [comment]
Perfect, I've removed the suggested comments. When you approve the changes it can be merged. Thx!

**Review by jacobduijzer — APPROVED** [review]

**Pull request merged**